core/src/main/java/software/amazon/awssdk/interceptor/ExecutionInterceptor.java

```diff
@@ -0,0 +1,297 @@
+import software.amazon.awssdk.interceptor.context.AfterMarshallingContext;
+import software.amazon.awssdk.interceptor.context.AfterTransmissionContext;
+import software.amazon.awssdk.interceptor.context.AfterUnmarshallingContext;
+import software.amazon.awssdk.interceptor.context.BeforeExecutionContext;
+import software.amazon.awssdk.interceptor.context.BeforeMarshallingContext;
+import software.amazon.awssdk.interceptor.context.BeforeTransmissionContext;
+import software.amazon.awssdk.interceptor.context.BeforeUnmarshallingContext;
+import software.amazon.awssdk.interceptor.context.FailedExecutionContext;
+import software.amazon.awssdk.retry.v2.RetryPolicy;
+
+/**
+ * An interceptor that is invoked during the execution lifecycle of a request/response (execution). This can be used to publish
+ * metrics, modify a request in-flight, debug request processing, view exceptions, etc. This interface exposes different methods
+ * for hooking into different parts of the lifecycle of an execution.
+ *
+ * <p>
+ * <b>Interceptor Hooks</b>
+ * Methods for a given interceptor are executed in a predictable order, each receiving the information that is known about the
+ * message so far as well as a {@link ExecutionAttributes} object for storing data that is specific to a particular execution.
+ * <ol>
+ *     <li>{@link #beforeExecution} - Read the request before it is modified by other interceptors.</li>
+ *     <li>{@link #modifyRequest} - Modify the request object before it is marshalled into an HTTP request.</li>
+ *     <li>{@link #beforeMarshalling} - Read the request that has potentially been modified by other request interceptors before
+ *     it is marshalled into an HTTP request.</li>
+ *     <li>{@link #afterMarshalling} - Read the HTTP request before it is modified by other inteceptors.</li>
+ *     <li>{@link #modifyHttpRequest} - Modify the HTTP request object before it is transmitted.</li>
+ *     <li>{@link #beforeTransmission} - Read the HTTP request that has potentially been modified by other request interceptors
+ *     before it is sent to the service.</li>
+ *     <li>{@link #afterTransmission} - Read the HTTP response before it is modified by other interceptors.</li>
+ *     <li>{@link #modifyHttpResponse} - Modify the HTTP response object before it is unmarshalled.</li>
+ *     <li>{@link #beforeUnmarshalling} - Read the HTTP response that has potentially been modified by other request interceptors
+ *     before it is unmarshalled.</li>
+ *     <li>{@link #afterUnmarshalling} - Read the response before it is modified by other interceptors.</li>
+ *     <li>{@link #modifyResponse} - Modify the response object before before it is returned to the client.</li>
+ *     <li>{@link #afterExecution} - Read the response that has potentially been modified by other request interceptors.</li>
+ * </ol>
+ * An additional {@link #onExecutionFailure} method is provided that is invoked if an execution fails at any point during the
```

Comment:
As you noticed, this is described below. Do you think it's relevant to include up here in the summary as well?